src/audio/pulseaudio/SDL_pulseaudio.c

```diff
@@ -682,19 +682,89 @@ static SDL_Thread *hotplug_thread = NULL;
 static void
 SinkInfoCallback(pa_context *c, const pa_sink_info *i, int is_last, void *data)
 {
+    SDL_AudioSpec spec;
     if (i) {
-        SDL_AddAudioDevice(SDL_FALSE, i->description, (void *) ((size_t) i->index+1));
+        spec.freq = i->sample_spec.rate;
+        spec.channels = i->sample_spec.channels;
+        switch (i->sample_spec.format) {
+        case PA_SAMPLE_U8:
+            spec.format = AUDIO_U8;
+            break;
+        case PA_SAMPLE_S16LE:
+            spec.format = AUDIO_S16LSB;
+            break;
+        case PA_SAMPLE_S16BE:
+            spec.format = AUDIO_S16MSB;
+            break;
+        case PA_SAMPLE_S32LE:
+            spec.format = AUDIO_S32LSB;
+            break;
+        case PA_SAMPLE_S32BE:
+            spec.format = AUDIO_S32MSB;
+            break;
+        case PA_SAMPLE_FLOAT32LE:
+            spec.format = AUDIO_F32LSB;
+            break;
+        case PA_SAMPLE_FLOAT32BE:
+            spec.format = AUDIO_F32MSB;
+            break;
+        default:
+            spec.format = 0;
+            break;
+        }
+        spec.silence = 0;
+        spec.samples = 0;
+        spec.size = 0;
+        spec.callback = NULL;
+        spec.userdata = NULL;
+
+        SDL_AddAudioDevice(SDL_FALSE, i->description, &spec, (void *) ((size_t) i->index+1));
     }
 }
 
 /* This is called when PulseAudio adds a capture ("source") device. */
 static void
 SourceInfoCallback(pa_context *c, const pa_source_info *i, int is_last, void *data)
 {
+    SDL_AudioSpec spec;
     if (i) {
         /* Skip "monitor" sources. These are just output from other sinks. */
         if (i->monitor_of_sink == PA_INVALID_INDEX) {
-            SDL_AddAudioDevice(SDL_TRUE, i->description, (void *) ((size_t) i->index+1));
+            spec.freq = i->sample_spec.rate;
+            spec.channels = i->sample_spec.channels;
+            switch (i->sample_spec.format) {
+            case PA_SAMPLE_U8:
+                spec.format = AUDIO_U8;
+                break;
+            case PA_SAMPLE_S16LE:
+                spec.format = AUDIO_S16LSB;
+                break;
+            case PA_SAMPLE_S16BE:
+                spec.format = AUDIO_S16MSB;
+                break;
+            case PA_SAMPLE_S32LE:
+                spec.format = AUDIO_S32LSB;
+                break;
+            case PA_SAMPLE_S32BE:
+                spec.format = AUDIO_S32MSB;
+                break;
+            case PA_SAMPLE_FLOAT32LE:
+                spec.format = AUDIO_F32LSB;
+                break;
+            case PA_SAMPLE_FLOAT32BE:
+                spec.format = AUDIO_F32MSB;
+                break;
+            default:
+                spec.format = 0;
+                break;
+            }
+            spec.silence = 0;
+            spec.samples = 0;
+            spec.size = 0;
+            spec.callback = NULL;
+            spec.userdata = NULL;
+
+            SDL_AddAudioDevice(SDL_TRUE, i->description, &spec, (void *) ((size_t) i->index+1));
```

Comment:
Can we move this to a common function instead of copy/pasting it between SourceInfoCallback and SinkInfoCallback?

Reply:
(The switch statement, I mean, which I failed to highlight properly.)